src/Analyzers/CSharp/Analyzers/MakeStructReadOnly/CSharpMakeStructReadOnlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp.MakeStructReadOnly;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+internal class CSharpMakeStructReadOnlyDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+{
+    public CSharpMakeStructReadOnlyDiagnosticAnalyzer()
+        : base(IDEDiagnosticIds.MakeStructReadOnlyDiagnosticId,
+               EnforceOnBuildValues.MakeStructReadOnly,
+               CSharpCodeStyleOptions.PreferDeconstructedVariableDeclaration,
+               new LocalizableResourceString(nameof(CSharpAnalyzersResources.Make_struct_readonly), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)),
+               new LocalizableResourceString(nameof(CSharpAnalyzersResources.Struct_can_be_made_readonly), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+    {
+    }
+
+    public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+        => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+    protected override void InitializeWorker(AnalysisContext context)
+        => context.RegisterSymbolStartAction(context =>
+        {
+            var compilation = context.Compilation;
+            if (compilation.LanguageVersion() < LanguageVersion.CSharp7_2)
+                return;
+
+            // First, see if this at least strongly looks like a struct that could be converted.
+            if (!IsCandidate(context, out var typeDeclaration, out var option))
+                return;
+
+            // Looks good.  However, we have to make sure that the struct has no code which actually overwrites 'this'
+
+            var writesToThis = false;
+            context.RegisterSyntaxNodeAction(context =>
+            {
+                var semanticModel = context.SemanticModel;
+                var thisExpression = (ThisExpressionSyntax)context.Node;
+                var cancellationToken = context.CancellationToken;
+                writesToThis = writesToThis || thisExpression.IsWrittenTo(semanticModel, cancellationToken);
+            }, SyntaxKind.ThisExpression);
+
+            context.RegisterSymbolEndAction(context =>
+            {
+                // if we wrote to 'this', then we cannot convert this struct.
+                if (writesToThis)
+                    return;
+
+                context.ReportDiagnostic(DiagnosticHelper.Create(
+                    Descriptor,
+                    typeDeclaration.Identifier.GetLocation(),
```

Comment:
```suggestion
                    context.Symbol.Locations[0],
```

Reply:
i want this specifically on the identifier.  Symbol.Locations doesn't specify where that is.

Reply:
@CyrusNajmabadi I believe `Symbol.Locations` will be the identifier.

Reply:
there is no guarantee of this.  i far prefer just grabbing the known syntax and placing the identifier exactly where i would like it to be :)